src/config.ts

```diff
@@ -333,34 +333,40 @@ export class KubeConfig {
         }
     }
 
-    private applyAuthorizationHeader(opts: request.Options | https.RequestOptions) {
+    private applyAuthentication(opts: request.Options | https.RequestOptions) {
         const user = this.getCurrentUser();
         if (!user) {
             return;
         }
-        let token: string | null = null;
 
-        KubeConfig.authenticators.forEach((authenticator: Authenticator) => {
+        let credentials: Credentials | null = null;
+
+        for (const authenticator of KubeConfig.authenticators) {
             if (authenticator.isAuthProvider(user)) {
-                token = authenticator.getToken(user);
+                credentials = authenticator.getCredentials(user);
             }
-        });
-
-        if (user.token) {
-            token = 'Bearer ' + user.token;
         }
 
+        // Apply bearer token, if provided
+        const token = user.token || (credentials && credentials.type === 'token' && credentials.token);
```

Comment:
Again I think all this checking logic could be switched to  `const token = user.token || credentials.token`. 

And then the same for the cert data too.

Reply:
Yeah unfortunately doesn't work directly, TypeScript will complain about referencing the missing attribute. But, I could instead work out different types of type guards (https://basarat.gitbooks.io/typescript/docs/types/typeGuard.html), maybe use an "is" function instead of the literals.

Reply:
This assumed that you put both the token and the client data in the same credential type/class. :) 

```js
export interface Credential {
    token?: string;
    clientCertificateData?: string;
    clientKeyData?: string;
```

In the current, we are guarding against by type casting/checking in the above we are being slightly more duck typing and assuming if it has a token. 

If you look at the code currently it supports passing both a token and a client cert (technically it's not in an else block but with the type checking that flow doesn't work or isn't tested)

It makes the code cleaner/simpler since it doesn't do any manipulation/conversions it just sets the token if present and sets the cert data if present and then passes that information upstream to the API's.

This is actually how the go-client does it https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/validation.go#L198 It makes it a bunch easier if the clients follow similar logic (since they get ported between all the languages)

But maybe @brendandburns can weigh in here too. I feel like both have pluses and minuses I just prefer the putting all the auth info into the same type (since we are literally doing `|` or types when ever we use it)

Reply:
Yeah I don't know, this line throws if the credential has more than one method: https://github.com/kubernetes/client-go/blob/7de88b14dcc8f8ebdc2953e776666387205cb592/tools/clientcmd/validation.go#L261

Seems we shouldn't allow both at least. The method can be re-written to be more analogous to the client-go code and to do more thorough validation, but overall the TS lib seems nowhere near parity in the first place. So I don't know, up to you. We're running off our own fork now anyway so no rush on my part.